Add json/protobuf/yaml fixtures

liggitt opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Adds roundtrip testing to/from persisted fixtures we can snapshot per release.

This will give us confidence bumps of proto libraries aren't breaking backwards compatibility with old serialized data.

**Special notes for your reviewer**:

Follow-ups:
- [x] include regeneration of fixtures for HEAD in `make update` - https://github.com/kubernetes/kubernetes/pull/78656
- [x] help remaining API-change PRs targeting 1.15 update fixtures (https://github.com/kubernetes/kubernetes/pull/78309#issuecomment-497445624)
  - https://github.com/kubernetes/kubernetes/pull/76968#issuecomment-498077997
  - https://github.com/kubernetes/kubernetes/pull/77595#issuecomment-498077818
- [ ] ~~add fixtures for apiextensions (CRD and ConversionReview types - fuzzing needs some special casing)~~ tracked in https://github.com/kubernetes/kubernetes/issues/79026
- [ ] ~~add fixtures for kube-aggregator (APIService types)~~ tracked in https://github.com/kubernetes/kubernetes/issues/79027
- [ ] ~~add fixtures for table output (PartialObjectMetadata, Table types - fuzzing needs some special casing)~~ tracked in https://github.com/kubernetes/kubernetes/issues/79028
- [ ] add to release process docs to snapshot HEAD fixtures to v1.x.0 dir for each release, e.g.:
    ```sh
    git checkout v1.15.0
    cp -fr ./staging/src/k8s.io/api/testdata/HEAD/ ./staging/src/k8s.io/api/testdata/v1.15.0
    git checkout master
    git add ./staging/src/k8s.io/api/testdata/v1.15.0
    git commit -m "Add API compatibility data for v1.15.0
    ```
- [ ] ~~Fix RawExtension normalization in yaml->json which causes DeepEqual failure~~ tracked in https://github.com/kubernetes/kubernetes/issues/73655

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig api-machinery
/sig architecture
/cc @smarterclayton

liggitt commented:
looks like the diffs in serialized proto between 1.13 and 1.14 and master are due to new optional but non-pointer tags added to API objects that are unconditionally written in proto serializations:

for example, this was the non-pointer Tenant field added to QuobyteVolumeSource:
```
--- FAIL: TestRoundTripExternalTypes/apps.v1beta1.Deployment (0.07s)
    --- FAIL: TestRoundTripExternalTypes/apps.v1beta1.Deployment/fixtures-v1.13.0 (0.03s)
        fixtures.go:195: proto data differed from fixture
        fixtures.go:205:   strings.Join({
              	... // 244 identical lines
              	`              4: "\347\256\220\351\270\2570w\310\263\306\242"`,
              	`              5: "\302\267\346\265\243l\350\223\222 "`,
            + 	`              6: ""`,
              	"            }",
              	"            22 {",
              	... // 157 identical lines
              	`            4: "\304\216D\350\262\270\350\264\205$\345\233\274\305\251J\345\243\2519\351\270\225\345\262\231"`,
              	`            5: "|"`,
            + 	`            6: ""`,
              	"          }",
              	"          10 {",
              	... // 370 identical lines
              	`            4: ""`,
              	"            5: \"\\312\\253!`\\350\\243\\263t\\350\\277\\215\\345\\206\\261\\311\\270\\302\\241\\306\\274\\304\\253\\310\\235\\310\\270\\304\\220\\307\\2655q\\304\\225\\306\\223\"",
            + 	`            6: ""`,
              	"          }",
              	"          10 {",
              	... // 239 identical lines
              }, "\n")
```

and this was the RemainingListItemCount field added to listmeta:
```
--- FAIL: TestRoundTripExternalTypes (1.67s)
    --- FAIL: TestRoundTripExternalTypes/.v1.Status (0.04s)
        --- FAIL: TestRoundTripExternalTypes/.v1.Status/fixtures-v1.14.0 (0.04s)
            fixtures.go:195: proto data differed from fixture
            fixtures.go:205: fixture vs reserialized:
                  strings.Join({
                  	... // 7 identical lines
                  	`    2: "8551505577692297499"`,
                  	`    3: ""`,
                + 	"    4: 0",
                  	"  }",
                  	"  2 {",
                  	... // 21 identical lines
                  }, "\n")
```

smarterclayton reviewed (changes requested):
Some structural things

liggitt committed:
Don't fuzz deprecated initializers field

liggitt committed:
Remove broken round trip test

liggitt committed:
Add compatibility tests for json/proto/yaml

liggitt committed:
Run roundtrip tests as subtests

liggitt committed:
Set GVK when decoding protobuf

liggitt committed:
Compatibility data: master

liggitt committed:
Compatibility data: v1.14.0

liggitt committed:
Generate bazel

liggitt committed:
Update expected ListOptions

        --- FAIL: TestCompatibility/core.v1.ListOptions/v1.14.0 (0.02s)
            compatibility.go:472: proto differs
            compatibility.go:474:   strings.Join({
                  	... // 10 identical lines
                  	"  7: 6780787122834727873",
                  	`  8: "5"`,
                + 	"  9: 0",
                  	"}",
                  	`3: ""`,
                  	... // 2 identical lines
                  }, "\n")

Caused by https://github.com/kubernetes/kubernetes/pull/74074/files#diff-eca3b8d856fa2e661f6da91b61de5e76R364

liggitt committed:
Update expected AdmissionReview

        --- FAIL: TestCompatibility/admission.k8s.io.v1beta1.AdmissionReview/v1.14.0 (0.02s)
            compatibility.go:460: json differs
            compatibility.go:461:   strings.Join({
                  	... // 31 identical lines
                  	`    "object": {"apiVersion":"example.com/v1","kind":"CustomType","spec":{"replicas":1},"status":{"available":1}},`,
                  	`    "oldObject": {"apiVersion":"example.com/v1","kind":"CustomType","spec":{"replicas":1},"status":{"available":1}},`,
                - 	`    "dryRun": true`,
                + 	`    "dryRun": true,`,
                + 	`    "options": null`,
                  	"  },",
                  	`  "response": {`,
                  	... // 33 identical lines
                  }, "\n")

            compatibility.go:466: yaml differs
            compatibility.go:467:   strings.Join({
                  	... // 23 identical lines
                  	"      available: 1",
                  	"  operation: 祈¡ıŵDz廔ȇ{sŊƏp饏姥呄鐊",
                + 	"  options: null",
                  	"  resource:",
                  	`    group: "5"`,
                  	... // 38 identical lines
                  }, "\n")

            compatibility.go:472: proto differs
            compatibility.go:474:   strings.Join({
                  	... // 37 identical lines
                  	"    }",
                  	"    11: 1",
                + 	`    12: ""`,
                + 	`    15: ""`,
                  	"  }",
                  	"  2 {",
                  	... // 36 identical lines
                  }, "\n")

Null `options` in json and additional proto tag 12 caused by https://github.com/kubernetes/kubernetes/pull/77563#discussion_r282803268

Proto tag 15 caused by https://github.com/kubernetes/kubernetes/pull/78135/files#diff-3bc4acaf71b6b50648da046287017c54R82

liggitt commented:
addressed comments required for initial merge. I want to get this in asap to get primary API types protected, and do the additional API dirs as follow-ups

liggitt commented:
/priority important-soon

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -407,6 +407,9 @@ func unmarshalToObject(typer runtime.ObjectTyper, creater runtime.ObjectCreater,
 	if err := proto.Unmarshal(data, pb); err != nil {
 		return nil, actual, err
 	}
+	if actual != nil {
```
Hrm.  This was deliberate, but we since started to argue more that clearing gvk in memory wasn't really necessary.

I'm wondering whether anyone would be impacted.

liggitt replied:
not being able to round trip without external fixup seemed bad

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go:
```diff
@@ -140,13 +140,14 @@ func RoundTripExternalTypes(t *testing.T, scheme *runtime.Scheme, codecFactory r
-		// should be setting these two fields accordingly when protobuf is passed as codec
-		// to roundTrip method.
-		roundTripSpecificKind(t, gvk, scheme, codecFactory, fuzzer, nonRoundTrippableTypes, true)
+		t.Run(gvk.Group+"."+gvk.Version+"."+gvk.Kind, func(t *testing.T) {
```
Finally

smarterclayton commented:
/lgtm
/approve

liggitt commented:
/hold

coordinating with existing API PRs in flight to pipeline merges before freeze

liggitt commented:
/sig testing

liggitt commented:
/hold cancel
/test pull-kubernetes-bazel-rest

merge queue is emptyish. will need to update this for a few API PRs that merged

liggitt commented:
/test pull-kubernetes-bazel-test

liggitt commented:
rebased/regenerated fixtures for master

liggitt commented:
retagging after rebase

liggitt commented:
/retest
